src/watchman_client.js

```diff
@@ -0,0 +1,376 @@
+                                       ? { watchmanBinaryPath: watchmanBinaryPath}
+                                       : {});
+
+      // reset the relevant local values
+
+      try {
+        client.capabilityCheck(
+          {optional: ['wildmatch', 'relative_root']},
+          (error, resp) => {
+            if (error) {
+              reject(error);
+            } else {
+              this._watchmanBinaryPath = watchmanBinaryPath;
+              this._wildmatch = resp.capabilities.wildmatch;
+              this._relative_root = resp.capabilities.relative_root;
+              this._client = client;
+
+              this._setupClientListeners();
+              resolve(this);
+            }
+          }
+        );
+      } catch (err) {
+        console.error("Client capabilityCheck threw error somehow: " + err);
+        // XXX Do something here to throw an error or reject?
+      }
+    });
+  }
+
+  return this._clientPromise;
+};
+
+/**
+ * Called from WatchmanWatcher (or this object during reconnect) to create
+ * a watcherInfo entry in our _watcherMap and issue a 'subscribe' to the
+ * watchman.Client, to be handled in this object.
+ */
+WatchmanClient.prototype.subscribe = function(watchmanWatcher, root) {
+
+  let watcherInfo = this._createWatcherInfo(watchmanWatcher);
+  let subscription = watcherInfo.subscription;
+  
+  return this._watch(subscription, root)
+    .then((resp) => this._clock(subscription))
+    .then((data) => this._subscribe(subscription))
+    .catch((error) => {
+      console.error("Caught error in subscribe for watchmanWatcher root " + root);
+      console.error(error);
+      return Promise.reject(error); // XXX is this right? Or do I just do 'reject(error)'?
+    });
+};
+
+/**
+ * Remove the information about a specific WatchmanWatcher.
+ * Once done, if no watchers are left, clear the local vars,
+ * which will end the connection to the watchman.Client, too.
+ */
+WatchmanClient.prototype.closeWatcher = function(watchmanWatcher) {
+  let watcherInfos = Object.values(this._watcherMap);
+  let numWatchers = watcherInfos.length;
+
+  if (numWatchers > 0) {
+    let watcherInfo;
+    
+    for (let info of watcherInfos) {
+      if (info.watchmanWatcher === watchmanWatcher) {
+        watcherInfo = info;
+        break;
+      }
+    }
+
+    if (watcherInfo) {
+      delete this._watcherMap[watcherInfo.subscription];
+
+      numWatchers--;
+
+      if (numWatchers === 0) {
+        this._clearLocalVars();
+      }
+    }
+  }
+};
+
+/**
+ * Clear out local state at the beginning and if we end up
+ * getting disconnected and try to reconnect.
+ */
+WatchmanClient.prototype._clearLocalVars = function() {
+  if (this._client) {
+    this._client.removeAllListeners();
+    this._client.end();
+  }
+
+  this._client = null;
+  this._clientPromise = null;
+  this._wildmatch = false;
+  this._relative_root = false;
+  this._subscriptionId = 1;
+  this._watcherMap = {};
```

Comment:
lets make this guy `Object.create(null)` as we are using it as a mutating lookup table (which Object.create(null) are designed for), and likely don't want to be susceptible to pollution of the global object prototype.

Reply:
Sorry, I don't understand what you mean here by either Object.create(null) or 'mutating lookup table'. Can you explain more? this._clearLocalVars is used twice - once during construction and once during reconnect after _onEnd.

Reply:
TL;DR for dictionaries, `Object.create(null)` is more appropriate.

Rationale: 

* it has no prototype to fall back to on misses
* it cannot be accidentally polluted  if some other library mucks with `Object.prototype`
* v8 fast paths `Object.create(null)` for behaviors such as deletion, adding, and missed lookups.